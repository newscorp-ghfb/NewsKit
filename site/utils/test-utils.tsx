/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme: any = renderToFragmentWithThemeFactory(
  docsThemeLight,
);

export const mockNewsKitComponents = (...components: string[]) => () => ({
  ...(jest.requireActual('newskit') as any),
  ...components.reduce((acc, comp) => {
    acc[comp] = ({children, ...props}: any) => (
      <div
        data-comp={`NewsKit ${comp}`}
        data-testid={comp}
        data-props={JSON.stringify(props, null, 2)}
      >
        {children}
      </div>
    );
    return acc;
  }, {} as Record<string, Function>),
});
