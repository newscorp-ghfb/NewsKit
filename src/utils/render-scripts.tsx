import React from 'react';
import {getSSRId} from './get-ssr-id';

interface ExternalScriptData {
  src: string;
  async?: boolean;
}

interface InlineScriptData {
  content: string;
}

type ScriptData = ExternalScriptData | InlineScriptData;
interface HelmetScriptProps {
  type: 'text/javascript';
  src?: string;
  async?: boolean;
  innerHTML?: string;
}

const isExternalScriptData = (
  scriptData: ScriptData,
): scriptData is ExternalScriptData => 'src' in scriptData;

export interface RenderScriptsReactHelmetProp {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reactHelmet?: React.ComponentType<{script?: Array<any>}>;
  nonce?: string;
}

export interface RenderScriptsProps extends RenderScriptsReactHelmetProp {
  scripts: ScriptData[];
}

const parseContent = (content: string) =>
  content.includes('__FUNC__')
    ? content
        .replace(
          // eslint-disable-next-line no-useless-escape
          /\"*__FUNC__\"*/g,
          '',
        )
        .replace(/\\n+/g, String.fromCharCode(10))
        .replace(/\\+/g, '')
    : content;

export const RenderScripts: React.FC<RenderScriptsProps> = ({
  scripts,
  reactHelmet: ReactHelmet,
  nonce,
}) => {
  if (ReactHelmet) {
    return (
      <ReactHelmet
        script={[
          ...scripts.map(scriptData => {
            const scriptProps = {
              type: 'text/javascript',
            } as HelmetScriptProps;

            if (isExternalScriptData(scriptData)) {
              scriptProps.src = scriptData.src;
              scriptProps.async = scriptData.async;
            } else {
              scriptProps.innerHTML = parseContent(scriptData.content);
            }
            return scriptProps;
          }),
        ]}
      />
    );
  }

  return (
    <>
      {scripts.map(scriptData => {
        const scriptProps = isExternalScriptData(scriptData)
          ? {
              key: scriptData.src,
              src: scriptData.src,
              async: scriptData.async,
            }
          : {
              key: getSSRId(),
              dangerouslySetInnerHTML: {
                __html: parseContent(scriptData.content),
              },
            };
        return <script type="text/javascript" {...scriptProps} nonce={nonce} />;
      })}
    </>
  );
};
