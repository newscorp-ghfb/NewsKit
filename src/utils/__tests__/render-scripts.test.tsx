import React from 'react';
import {renderToFragment} from '../../test/test-utils';
import {RenderScripts, RenderScriptsProps} from '../render-scripts';

describe('RenderScripts', () => {
  const scripts: RenderScriptsProps['scripts'] = [
    {
      src: 'https://example.com/my-javascript-file.js',
      content: '',
    },
    {
      content: `(function () { alert('some injected script') })();`,
    },
    {
      src: 'https://example.com/my-javascript-file2.js',
      content: '',
    },
    {
      content: `(function () { alert('some more injected script') })();`,
    },
  ];

  const scriptsWithFunctions: RenderScriptsProps['scripts'] = [
    {
      content: `{
            "onMessageReady": "__FUNC__ () => {\n            console.log('onMessageReady');\n          }__FUNC__",
            "onConsentReady": "__FUNC__function onConsentReady(consentUUID, euconsent) {\n            console.log('onConsentReady');\n            }__FUNC__"
          }
        }
      }`,
    },
  ];

  test('renders inline and external scripts as script tags', () => {
    const fragment = renderToFragment(<RenderScripts scripts={scripts} />);
    expect(fragment).toMatchSnapshot();
  });

  test('renders scripts and correctly parses objects with events', () => {
    const fragment = renderToFragment(
      <RenderScripts scripts={scriptsWithFunctions} />,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('renders with nonce', () => {
    const fragment = renderToFragment(
      <RenderScripts scripts={scriptsWithFunctions} nonce="we32s235" />,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('renders inline and external scripts via ReactHelmet', () => {
    const MockReactHelmet: React.FC = props => (
      <div>MockReactHelmet props: {JSON.stringify(props, null, 2)}</div>
    );

    const fragment = renderToFragment(
      <RenderScripts scripts={scripts} reactHelmet={MockReactHelmet} />,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('renders scripts and correctly parses objects with events via ReactHelmet', () => {
    const MockReactHelmet: React.FC = props => (
      <div>MockReactHelmet props: {JSON.stringify(props, null, 2)}</div>
    );

    const fragment = renderToFragment(
      <RenderScripts
        scripts={scriptsWithFunctions}
        reactHelmet={MockReactHelmet}
      />,
    );
    expect(fragment).toMatchSnapshot();
  });
});
