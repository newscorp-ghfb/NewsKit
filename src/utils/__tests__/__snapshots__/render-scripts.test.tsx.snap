// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderScripts renders inline and external scripts as script tags 1`] = `
<DocumentFragment>
  <script
    src="https://example.com/my-javascript-file.js"
    type="text/javascript"
  />
  <script
    type="text/javascript"
  >
    (function () { alert('some injected script') })();
  </script>
  <script
    src="https://example.com/my-javascript-file2.js"
    type="text/javascript"
  />
  <script
    type="text/javascript"
  >
    (function () { alert('some more injected script') })();
  </script>
</DocumentFragment>
`;

exports[`RenderScripts renders inline and external scripts via ReactHelmet 1`] = `
<DocumentFragment>
  <div>
    MockReactHelmet props: {
  "script": [
    {
      "type": "text/javascript",
      "src": "https://example.com/my-javascript-file.js"
    },
    {
      "type": "text/javascript",
      "innerHTML": "(function () { alert('some injected script') })();"
    },
    {
      "type": "text/javascript",
      "src": "https://example.com/my-javascript-file2.js"
    },
    {
      "type": "text/javascript",
      "innerHTML": "(function () { alert('some more injected script') })();"
    }
  ]
}
  </div>
</DocumentFragment>
`;

exports[`RenderScripts renders scripts and correctly parses objects with events 1`] = `
<DocumentFragment>
  <script
    type="text/javascript"
  >
    {
            "onMessageReady":  () =&gt; {
            console.log('onMessageReady');
          },
            "onConsentReady": function onConsentReady(consentUUID, euconsent) {
            console.log('onConsentReady');
            }
          }
        }
      }
  </script>
</DocumentFragment>
`;

exports[`RenderScripts renders scripts and correctly parses objects with events via ReactHelmet 1`] = `
<DocumentFragment>
  <div>
    MockReactHelmet props: {
  "script": [
    {
      "type": "text/javascript",
      "innerHTML": "{\\n            \\"onMessageReady\\":  () =&gt; {\\n            console.log('onMessageReady');\\n          },\\n            \\"onConsentReady\\": function onConsentReady(consentUUID, euconsent) {\\n            console.log('onConsentReady');\\n            }\\n          }\\n        }\\n      }"
    }
  ]
}
  </div>
</DocumentFragment>
`;

exports[`RenderScripts renders with nonce 1`] = `
<DocumentFragment>
  <script
    nonce="we32s235"
    type="text/javascript"
  >
    {
            "onMessageReady":  () =&gt; {
            console.log('onMessageReady');
          },
            "onConsentReady": function onConsentReady(consentUUID, euconsent) {
            console.log('onConsentReady');
            }
          }
        }
      }
  </script>
</DocumentFragment>
`;
